File: src/main/java/io/github/emcw/utils/GsonUtil.java
package io.github.emcw.utils;

import io.github.emcw.adapters.ColorAdapter;
import io.github.emcw.adapters.DurationAdapter;

import com.google.gson.*;
import com.google.gson.reflect.TypeToken;

import lombok.AccessLevel;
import lombok.Getter;
import lombok.NoArgsConstructor;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.awt.*;
import java.lang.reflect.Type;
import java.time.Duration;
import java.util.*;
import java.util.List;
import java.util.concurrent.ConcurrentHashMap;

import java.util.stream.Collectors;
import java.util.stream.Stream;

import static java.util.Objects.requireNonNull;

@SuppressWarnings("unchecked, unused")
@NoArgsConstructor(access = AccessLevel.PRIVATE)
public class GsonUtil {
    @Getter private static final Gson GSON = new GsonBuilder()
        .registerTypeAdapter(Color.class, new ColorAdapter())
        .registerTypeAdapter(Duration.class, new DurationAdapter())
        .setPrettyPrinting()
        .serializeNulls()
        .create();

    public static String serialize(Object obj) {
        return GSON.toJson(obj, getType(obj));
    }

    public static <T> Type getType(@NotNull T obj) {
        return TypeToken.get(obj.getClass()).getType();
    }

    public static <T> Type getType(Class<T> clazz) {
        return TypeToken.getParameterized(clazz).getType();
    }

    public static <T> T deserialize(String str, Class<T> c) {
        return GSON.fromJson(str, c);
    }

    public static <T> T deserialize(String str, Type type) {
        return GSON.fromJson(str, type);
    }

    public static <T> T deserialize(JsonElement el, Type type) {
        return GSON.fromJson(el, type);
    }

    public static <T> T convert(Object obj, Class<T> clazz) {
        return deserialize(serialize(obj), clazz);
    }

    public static <T extends JsonElement> T asTree(Object input) {
        JsonElement tree = getGSON().toJsonTree(input);
        return tree.isJsonObject() ? (T) tree.getAsJsonObject() : (T) tree;
    }

    public static <T> List<T> toList(Object obj) {
       return convert(obj, List.class);
    }

    public static <T> @NotNull JsonArray mapToArr(@NotNull Map<String, T> map) {
        JsonArray arr = new JsonArray();
        map.values().forEach(v -> arr.add(asTree(v)));

        return arr;
    }

    public static <T> @NotNull Map<String, T> arrToMap(@NotNull JsonArray arr, String key) {
        ConcurrentHashMap<String, T> map = new ConcurrentHashMap<>();
        arrAsStream(arr).forEach(el -> {
            JsonObject obj = el.getAsJsonObject();
            String k = keyAsStr(obj, key);
            if (k == null) return;
            
            map.put(k, deserialize(el, getType(el)));
        });

        return map;
    }

    public static int[] arrToIntArr(@NotNull JsonArray arr) {
        return convert(arr, int[].class);
    }

    public static Stream<JsonElement> arrAsStream(@NotNull JsonArray arr) {
        return arr.asList().parallelStream();
    }

    public static Stream<Map.Entry<String, JsonElement>> streamEntries(@NotNull JsonObject o) {
        return streamEntries(o.asMap());
    }

    public static <T> Stream<Map.Entry<String, T>> streamEntries(@NotNull Map<String, T> o) {
        return o.entrySet().parallelStream();
    }

    public static <T> Stream<T> streamValues(@NotNull Map<String, T> o) {
        return o.values().parallelStream();
    }

    public static Map<String, JsonObject> intersection(JsonArray arr, JsonArray arr2) {
        return arrAsStream(arr).flatMap(obj -> arrAsStream(arr2)
                .map(JsonElement::getAsJsonObject)
                .filter(obj2 -> Objects.equals(member(obj2, "name"), obj))
        ).collect(Collectors.toMap(obj -> keyAsStr(obj, "name"), obj -> obj));
    }

//    public static Map<String, Player> difference(JsonArray ops, JsonArray residents) {
//        return difference(ops, residents, "name");
//    }
//
//    public static Map<String, Player> difference(JsonArray ops, JsonArray residents, String key) {
//        Set<String> names = arrAsStream(residents).filter(Objects::nonNull)
//                .map(res -> keyAsStr(res.getAsJsonObject(), key))
//                .collect(Collectors.toSet());
//
//        Type playerListType = new TypeToken<List<Player>>(){}.getType();
//        List<Player> playerList = deserialize(serialize(ops), playerListType);
//
//        return playerList.parallelStream()
//                .filter(Objects::nonNull)
//                .filter(op -> !names.contains(op.getName()))
//                .collect(Collectors.toMap(BaseEntity::getName, Function.identity()));
//    }

//    public static <T> JsonObject valueAsObj(Map.Entry<String, T> entry) {
//        return (JsonObject) entryVal(entry);
//    }
//
//    static <T> T entryVal(@NotNull Map.Entry<String, T> entry) {
//        return entry.getValue();
//    }

    static JsonElement member(JsonObject o, String k) {
        return o == null ? null : o.get(k);
    }

    static boolean isNull(JsonElement el) {
        return el == JsonNull.INSTANCE || el == null;
    }

    public static @Nullable JsonObject keyAsObj(JsonObject o, String k) {
        JsonElement key = member(o, k);
        JsonObject result = new JsonObject();

        key.getAsJsonArray().forEach(el -> result.add(k, el));

        return isNull(key) ? null : result;
    }

    @Nullable
    public static Boolean keyAsBool(JsonObject o, String k) {
        JsonElement key = member(o, k);
        return isNull(key) ? null : key.getAsBoolean();
    }

    @Nullable
    public static Integer keyAsInt(JsonObject o, String k) {
        JsonElement key = member(o, k);
        return isNull(key) ? null : key.getAsInt();
    }

    @Nullable
    public static String keyAsStr(JsonObject o, String k) {
        JsonElement key = member(o, k);
        return isNull(key) ? null : key.getAsString();
    }

    @Nullable
    public static Double keyAsDouble(JsonObject o, String k) {
        JsonElement key = member(o, k);
        return isNull(key) ? null : key.getAsDouble();
    }

    public static JsonArray keyAsArr(JsonObject obj, String key) {
        JsonArray arr = new JsonArray();

        try { arr = requireNonNull(member(obj, key)).getAsJsonArray(); }
        catch (IllegalStateException e) {
            arr.add(obj.get(key));
        }

        return arr;
    }

    public static @NotNull JsonObject mapToObj(@NotNull Map<String, JsonObject> map) {
        JsonObject obj = new JsonObject();
        map.forEach(obj::add);
        return obj;
    }

    public static JsonArray arrAsJsonArray(Object[] objects) {
        return asTree(List.of(objects)).getAsJsonArray();
    }
}

File: src/main/kotlin/io/github/emcw/utils/GsonUtil.kt
package io.github.emcw.utils

import com.google.gson.*
import com.google.gson.reflect.TypeToken
import io.github.emcw.adapters.ColorAdapter
import io.github.emcw.adapters.DurationAdapter
import io.github.emcw.map.entities.BaseEntity
import io.github.emcw.map.entities.Player
import lombok.AccessLevel
import lombok.Getter
import lombok.NoArgsConstructor
import java.awt.Color
import java.lang.reflect.Type
import java.time.Duration
import java.util.*
import java.util.concurrent.ConcurrentHashMap
import java.util.function.Consumer
import java.util.function.Function
import java.util.stream.Collectors
import java.util.stream.Stream

@NoArgsConstructor(access = AccessLevel.PRIVATE)
object GsonUtil {
    @Getter
    private val GSON = GsonBuilder()
        .registerTypeAdapter(Color::class.java, ColorAdapter())
        .registerTypeAdapter(Duration::class.java, DurationAdapter())
        .setPrettyPrinting().create()

    @JvmStatic
    fun <T> serialize(obj: Any?): String {
        return GSON.toJson(obj, getType(obj))
    }

    fun <T> getType(obj: T): Type {
        return TypeToken.get(obj.javaClass).getType()
    }

    fun <T> getType(clazz: Class<T>?): Type {
        return TypeToken.getParameterized(clazz).type
    }

    fun <T> deserialize(str: String?, c: Class<T>?): T {
        return GSON.fromJson(str, c)
    }

    fun <T> deserialize(str: String?, type: Type?): T {
        return GSON.fromJson(str, type)
    }

    fun <T> deserialize(el: JsonElement?, type: Type?): T {
        return GSON.fromJson(el, type)
    }

    fun <T> convert(obj: Any?, clazz: Class<T>?): T {
        return deserialize(serialize<Any>(obj), clazz)
    }

    fun <T : JsonElement?> asTree(input: Any?): T {
        val tree: JsonElement = GsonUtil.getGSON().toJsonTree(input)
        return if (tree.isJsonObject) tree.asJsonObject as T else tree as T
    }

    fun <T> toList(obj: Any?): List<T> {
        return convert<List<*>>(obj, MutableList::class.java)
    }

    fun <T> mapToArr(map: Map<String?, T>): JsonArray {
        val arr = JsonArray()
        map.values.forEach(Consumer { v: T -> arr.add(asTree<JsonElement>(v)) })
        return arr
    }

    fun <T> arrToMap(arr: JsonArray, key: String?): Map<String, T> {
        val map = ConcurrentHashMap<String, T>()
        arrAsStream(arr).forEach { el: JsonElement ->
            val obj = el.asJsonObject
            val k = keyAsStr(obj, key) ?: return@forEach
            map[k] = deserialize(el, getType(el))
        }
        return map
    }

    fun arrToIntArr(arr: JsonArray): IntArray {
        return convert(arr, IntArray::class.java)
    }

    fun arrFromStrArr(obj: Array<String?>): JsonArray {
        val arr = JsonArray()
        for (value in obj) {
            arr.add(deserialize(value, JsonElement::class.java))
        }
        return arr
    }

    fun strArrAsStream(arr: Array<out String>): Stream<String> {
        return Stream.of(*arr).toList().parallelStream()
    }

    fun arrAsStream(arr: JsonArray): Stream<JsonElement> {
        return arr.asList().parallelStream()
    }

    fun streamEntries(o: JsonObject): Stream<Map.Entry<String?, JsonElement>> {
        return streamEntries(o.asMap())
    }

    fun <T> streamEntries(o: Map<String?, T>): Stream<Map.Entry<String?, T>> {
        return o.entries.parallelStream()
    }

    fun <T> streamValues(o: Map<String?, T>): Stream<T> {
        return o.values.parallelStream()
    }

    fun intersection(arr: JsonArray, arr2: JsonArray): Map<String?, JsonObject?> {
        return arrAsStream(arr).flatMap { obj: JsonElement? -> arrAsStream(arr2)
            .map { el: JsonElement -> el.asJsonObject }
            .filter { obj2: JsonObject? -> member(obj2, "name") == obj }
        }.collect(Collectors.toMap(
            { obj: JsonObject? -> keyAsStr(obj, "name") },
            { obj: JsonObject? -> obj })
        )
    }

    @JvmOverloads
    fun difference(ops: JsonArray?, residents: JsonArray, key: String? = "name"): Map<String, Player> {
        val names = arrAsStream(residents).filter { obj: JsonElement? -> Objects.nonNull(obj) }
            .map { res: JsonElement -> keyAsStr(res.asJsonObject, key) }
            .collect(Collectors.toSet())
        val playerListType = object : TypeToken<List<Player?>?>() {}.type
        val playerList = deserialize<List<Player>>(serialize<Any>(ops), playerListType)
        return playerList.parallelStream()
            .filter { obj: Player? -> Objects.nonNull(obj) }
            .filter { op: Player -> !names.contains(op.name) }
            .collect(Collectors.toMap(BaseEntity::name, Function.identity()))
    }

    fun <T> valueAsObj(entry: Map.Entry<String?, T>): JsonObject {
        return entryVal(entry) as JsonObject
    }

    fun <T> entryVal(entry: Map.Entry<String?, T>): T {
        return entry.value
    }

    fun member(o: JsonObject?, k: String?): JsonElement? {
        return o?.get(k)
    }

    fun isNull(el: JsonElement?): Boolean {
        return el === JsonNull.INSTANCE || el == null
    }

    fun keyAsObj(o: JsonObject?, k: String?): JsonObject? {
        val key = member(o, k)
        val result = JsonObject()
        key!!.asJsonArray.forEach(Consumer { el: JsonElement? -> result.add(k, el) })
        return if (isNull(key)) null else result
    }

    fun keyAsBool(o: JsonObject?, k: String?): Boolean {
        val key = member(o, k)
        return key != null && key.asBoolean
    }

    fun keyAsInt(o: JsonObject?, k: String?): Int? {
        val key = member(o, k)
        return if (isNull(key)) null else key!!.asInt
    }

    fun keyAsStr(o: JsonObject?, k: String?): String? {
        val key = member(o, k)
        return if (isNull(key)) null else key!!.asString
    }

    fun keyAsArr(obj: JsonObject, key: String?): JsonArray {
        var arr = JsonArray()
        try {
            arr = Objects.requireNonNull(member(obj, key)).asJsonArray
        } catch (e: IllegalStateException) {
            arr.add(obj[key])
        }
        return arr
    }

    fun mapToObj(map: Map<String?, JsonObject?>): JsonObject {
        val obj = JsonObject()
        map.forEach { (property: String?, value: JsonObject?) -> obj.add(property, value) }
        return obj
    }
}